Indicators api

akhenry opened the pull request:
In-progress PR for the Indicators API. For a simple indicator (90% of cases) it works like so:

```javascript
//var myIndicator = openmct.indicators.create(1000); // With optional priority 
var myIndicator = openmct.indicators.create();
myIndicator.text("Hello World!");
myIndicator.iconClass("icon-info");
```
In this scenario, Open MCT mostly manages how the indicator looks (though css classes can be applied with the appropriate function calls), and is completely responsible for updating the UI when the setter functions are called. The behavior of the indicator is left to external code under the control of the developer, with the setter functions used to update its onscreen appearance. See the updated [Follow Indicator](https://github.com/nasa/openmct/pull/1837/commits/081cbd3a1d55265c2cf08157c0013c61f7528db1) and [URL Status Indicator](https://github.com/nasa/openmct/pull/1837/commits/69ea40dcadeb1fdcc98265de69291c78f8478a4d) for examples of its use.

For a completely custom indicator you can pass a function into the call to `openmct.indicators.create(function)`. The only requirement for the function you provide is that it must return a DOM node for insertion alongside the other indicators. In this mode, no indicator object is returned, and the caller is entirely responsible for the look, behavior, and ongoing updating of the indicator.
eg.

```javascript
//openmct.indicators.create(1000, function () { // With optional priority
openmct.indicators.create(function () {
    var domNode = document.createElement('div');
    domNode.innerText = new Date().toString();
    setInterval(function () {
        domNode.innerText = new Date().toString();
    }, 1000);
    return domNode;
});
```

Looking for comments on the approach taken here.

To do
* [x] Allow definition of priorities for indicators
* [x] Wrap nodes returned by custom display functions. (Currently indicators MUST have certain classes attached otherwise they won't display properly).
* [x] API Documentation
* [x] Code cleanup
* [x] Tidy up commits
* [ ] ~~Convert default indicator to Vue component? ([Currently using vanilla JS](https://github.com/nasa/openmct/blob/indicators-api/src/api/indicators/Indicator.js#L84)).~~ Deferring for now. Not really sure it makes sense to do this anyway, it would end up being more code than the current implementation I suspect.
* [x] Unit tests
* [x] JSHint / JSCS

akhenry commented:
@VWoeltjen @larkin I'll hold off on the last three items until we've discussed what I've done so far.

akhenry commented:
Fixed an issue with the URL status indicator introduced by refactoring.

VWoeltjen commented:
> Convert default indicator to Vue component?

I'll suggest to defer this to a future issue, esp. if the tests you plan to write exercise the existing implementation in a way that will detect breakages from that change

VWoeltjen commented:
A few notes on the API in the current changeset. Only took a skim through the implementation, but looks good; liked seeing lots of Angularisms in red.

- `create` seems inconsistent with our other registration methods/naming. Either `addIndicator` or `register` would be more like what you've done elsewhere
- Returning a DOM node for the factory form of registration is also inconsistent; elsewhere we use a `View` interface with a `show(container)` and `destroy()`. We don't ever destroy indicators _currently_, but baking that into the API seems unwise and at minimum the consistency would be nice.
- To express the default form of create, I'd expose an `openmct.indicators.IndicatorView` which has all those API methods exposed, and implements the `View` interface.
- Including priority as an argument also feels funny; it's different from elsewhere and not intuitive what the argument means. I'd do it as a second `options` argument, or use a form like `{ view: new IndicatorView(), priority: 2 }` if that's cleaner.

akhenry commented:
@VWoeltjen Thanks for the feedback. I don't disagree with anything you've said, and there are certainly some departures from the rest of our API. I'll try to explain the thinking behind some of these decisions.

The aim was to make the 90% case as simple as possible, so that with a simple function call with no parameters, a new indicator can be added to the system. It can then be customized imperatively with calls to the returned interface. The idea here is that you could discover how to use it from the debugger, or from glancing at the code, without ever having to refer to documentation. Given the approach that I used, the verbs "add" and "register" didn't feel right (in the typical case we're not providing anything to add or register, unlike with views etc.). Requiring options objects, and demanding conformance with an interface require a fairly detailed reading of the documentation and aren't technically necessary in the fairly simple case of indicators.

Returning a DOM element is inconsistent with the view API, but I think indicators are conceptually a little different from views, and returning a DOM node gives the calling code more flexibility about placement of the indicator (which is important when we have many of them, with different priorities). That said, indicators could be rendered into containing divs under the control of the framework, so my argument here isn't very strong 😃 .

Destruction of indicators is big problem with my architecture. I didn't consider it because we don't ever destroy indicators, but you're right, it would be a mistake to design an API that makes it difficult to support destruction in future. Let me have a think about it...

VWoeltjen commented:
FWIW it could be a pretty modest change, e.g.:

```
openmct.indicators.create(3, function () {
    // create, activate, return DOM element
}, function destroy(domElement) {
    // hypothetically, for parity, deactivate a domElement
});
```

Could become:

```
openmct.indicators.addIndicator({
    priority: 3,
    show: function (container) {
        // create, activate, insert DOM element
    },
    destroy: function () {
        // deactivate DOM element
    }
);
```

Which is more verbose. Good for readability, but it does make things less concise.

That's the 10% case, as you point out. Maybe `create` could remain as-is for the 90% case? Excepting that `{ priority: 0 }` would be more readable/extensible as an argument, IMO. I think handling both of these cases from the same method is part of why it feels less intuitive to me.

akhenry commented:
Yeah, on balance an options object is going to be more readable than parameters. I'll make the change.

We also have to consider the destruction event for the default indicator. We could support an optional `destroy` function even in the absence of a `show` function.

larkin commented on src/api/indicators/Indicator.js:
```diff
@@ -0,0 +1,102 @@
+            }
+        }
+
+        Indicator.defaultDisplayFunction = function () {
```
is this supposed to be on prototype?

akhenry replied:
It's not, Indicator instances are a public interface, I don't want to expose anything that shouldn't be called by developers in our public API.

larkin commented:
Good stuff here.  My thoughts:

1. I think we should drop priority unless necessary.
    * setting priority via order of registration is "good enough" for me.
    * easy to add later...
2. Indicators ***should not*** inherit the full view interface.  ***They don't need it.***
     * ...Provided that we agree an indicator instance is an object:
         * an indicator instance must expose an element attribute that the platform will attach in the correct place.
         * we can add lifecycle methods at a later date if we need them. 
         * we can add priority at a later date if we need it.
3. I like the 90% case of overloading create, but I think the side effects are confusing
    * replace `create` with `register` or `add`.
    * as @VWoeltjen suggested, expose `SimpleIndicator` for simple out of the box behavior.


```javascript
// simple case, from above.
var myIndicator = openmct.indicators.SimpleIndicator();
myIndicator.text('Hello World!');
openmct.indicators.register(myIndicator);

// simple custom indicator:
var myIndicator = {
    element: document.createElement('div');
}
myIndicator.element.innerHTML = 'Hello World!';
openmct.indicators.register(myIndicator);

// more complex indicator:
class CurrentClockIndicator {
    constructor (openmct) {
        this.element = document.createElement('div');
        this.element.innerHTML = 'Hello World!';
        this.followClock = this.followClock.bind(this);
        this.onTick = this.onTick.bind(this);
        openmct.time.on('clock', this.followClock);
        this.followClock(openmct.time.clock());
    }
    followClock (clock) {
        if (this.clock) {
            this.clock.off('tick', this.onTick);
        }
        if (clock) {
            clock.on('tick', this.onTick);
        }
        this.clock = clock;
    }
    onTick (tick) {
        this.element.innerHTML = tick;
    }
}

var clockIndicator = new CurrentClockIndicator(openmct);
openmct.indicators.register(clockIndicator);
```

larkin commented on platform/features/clock/src/indicators/FollowIndicator.js:
```diff
@@ -20,38 +20,31 @@
-         */
-        function FollowIndicator(timerService) {
-            this.timerService = timerService;
+    function setIndicatorStatus(indicator, timer) {
```
IMO just stick this in the closure instead of binding it.

larkin commented:
I don't like the profusion of `iconClass`, `statusClass`; would rather abstract away CSS classes and enforce a standard set of UI concepts for the simple indicator, e.g. `indicator.icon('database'); indicator.status('warning');`  If they want to use different styles they can either extend SimpleIndicator or they can implement their own custom indicator.

akhenry commented:
Thanks for the reviews guys, some great suggestions. The proposed changes  I think are reasonably reconcilable, and are not a massive code delta from what I've implemented. I propose the following changes

* Expose a SimpleIndicator for the 90% case, which is then `.add`ed  via the API. This makes the 90% case a two-step process, rather than a single step, but it does make simple and custom indicators more consistent. 
* Pass an `object` into `.add()`. An object gives us more options in future. It can be used to add lifecycle events (like destroy) later, if necessary.
* Rename `.create()` to `.add`. I prefer `.add` to `.register`, because it's shorter and slightly more consistent with the rest of the new API. `.addIndicator` feels a little verbose, and a little redundant. So the call would be `openmct.indicators.add(indicator)`
* The object passed to `.add()` can optionally expose a `.element` attribute.
* Remove priority __for now__. The down side to this is that we lose a capability that we currently have in Indicators, and we may see some reordering of indicators (which we can potentially deal with via registration reordering). I wonder how big of a deal that really is though. Priorities don't give you 100% control over positioning of indicators anyway. If you have multiple indicators with priority Number.POSITIVE_INFINITY, then ordering comes down to order of registration anyway.

Where I would push back a bit is -
> would rather abstract away CSS classes and enforce a standard set of UI concepts for the simple indicator, e.g. indicator.icon('database'); 
My feeling is that this obfuscates things a bit. To me it feels like an unnecessary abstraction, and we're forcing the user to find out what an icon and a status are in the Open MCT world. It also means that the user has to write a completely custom indicator if they want to use something outside of our icon set. Also, the class names are referenced in the style-guide, so there's no additional documentation needed.

akhenry commented:
This is ready for final review and merge now. I've addressed the items we discussed, and added some test specs.

Note that build will currently fail due to jshint complaining about some late function definitions. I have created a PR that allows late definition of functions - https://github.com/nasa/openmct/pull/1887 . Once that's merged I'll rebase and this build will no longer fail.

## Author Checklist

* Changes address original issue? Y
* Unit tests included and/or updated with changes? Y
* Command line build passes? Y
* Changes have been smoke-tested? Y

VWoeltjen commented:
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

charlesh88 commented:
@akhenry Changes I had made in relation to this are in https://github.com/nasa/openmct/issues/1894.

larkin commented:
@akhenry @charlesh88 What is the status on this PR?

akhenry commented:
@larkin Aside from the new merge conflict, I think we're good to go. @charlesh88 you're done on your end?

Andrew

akhenry commented:
@larkin Aside from the new merge conflict, I think we're good to go. @charlesh88 you're done on your end?

larkin commented:
@akhenry any reason we deviated from `cssClass` to `iconClass` here?  We've standardized on that for some time.

charlesh88 commented:
@larkin Reason is that I removed a DOM element that was specifically for the icon, and do everything in the cssClass now - icon, status, etc.

larkin commented:
@charlesh88 right which is why I'd expect it to be cssClass, not iconClass.

akhenry commented:
The SimpleIndicator has an `iconClass` and a `statusClass`, and does not define a `cssClass`. The thinking behind this is that it's public API, and separating these things provides a semantic aid to developers. This is how you set the icon class, and this is how you set the warning class. This also makes it a little easier to have a  sensible defaults that can be progressively customized. Overriding a default icon is a little harder with a generic 'cssClass' function.

The URLIndicator was, I guess, changed for consistency, and for clarity.

larkin commented on src/api/indicators/IndicatorAPI.js:
```diff
@@ -0,0 +1,130 @@
+        this.openmct = openmct;
+        this.indicatorElements = [];
+        this.promiseForAllElements =
+            fetchLegacyIndicators.call(this)
```
Implement legacy indicator support in a separate file to separate concerns, plus that makes it trivial to remove legacy support later.

larkin commented on src/api/indicators/SimpleIndicator.js:
```diff
@@ -0,0 +1,96 @@
+            return this.descriptionValue;
+        };
+
+        SimpleIndicator.prototype.iconClass = function (iconClass) {
```
This is a nice middle road; I consider there to be two other options for this interface but I think this one is "just right".  

* isolate from style guide and simplify, i.e. predefine a set of icons/statuses and have them call `icon("clock"); status("connected");`, so they don't have to read style guide.
* tie more closely to style guide, have them set a single `cssClass` and then force them to understand that status classes can be combined with icon classes.

to be very clear: this is a comment, not a requested change.

larkin commented on src/plugins/URLIndicatorPlugin/URLIndicatorSpec.js:
```diff
@@ -21,138 +21,166 @@
+                return returned;
+            }
+
+            function mockInterval(interval) {
```
We are frugal developers, we speak the jasmine language, we should not introduce new words into the language when they are not necessary because then we have to teach those words to everyone.  Especially when the function name doesn't even match what it does (trigger an interval).

larkin commented on src/plugins/URLIndicatorPlugin/URLIndicatorSpec.js:
```diff
@@ -21,138 +21,166 @@
+                mockHttpRequestFunction
+                    .andReturn(Promise.reject().then(function () {
+                        returned = true;
+                        //Throw error to ensure chained catch is invoked
```
this comment is invalid?

larkin commented on src/plugins/URLIndicatorPlugin/URLIndicator.js:
```diff
@@ -21,77 +21,95 @@
+                .catch(this.handleError);
+        };
+
+        URLIndicator.prototype.get = function (url) {
```
You should just spy on $.ajax instead of exposing a method purely to support testing; this also makes the test unnecessarily asynchronous.

akhenry replied:
This was not done to support testing. I've made [my feelings](https://github.com/nasa/openmct/pull/1870#issuecomment-387175530) clear about writing code just to support tests. It was done to wrap the asynchronous behavior of this class in the promises model for consistency with the rest of our codebase and with modern HTTP APIs (eg. fetch). That said, it's extra code for marginal benefit here, so I'll just use the callbacks.

akhenry commented:
> @akhenry when you got to IndicatorAPI.js, you said "well, let me use-before-define a bunch of functions to hide all that complexity," when in reality that complexity should have been separated out into a runs-service which transforms and registers legacy indicators, and a spec for that run-service. That will cleanly separate code and would present an easier migration path when we remove legacy indicators.

For posterity's sake, these functions were integrated into `IndicatorAPI` because a complete list of indicators managed by the API cannot be provided until legacy indicators have been resolved, which happens asynchronously. This was addressed by exposing a promise that guarantees that the Indicators API will only return the list of indicators when all of the legacy ones have been provided. This unfortunately required the legacy code to be baked into the Indicator API to some extent. Furthermore, it was implemented as non-member functions to avoid exposing private legacy functions on public API.

I have now externalized the support for legacy indicators into a separate plugin, and removed the explicit guarantee.

akhenry commented on src/api/indicators/IndicatorAPISpec.js:
```diff
@@ -0,0 +1,221 @@
+
+            });
+
+            describe("The simple indicator", function () {
```
I'd like to push back a bit on this point. The SimpleIndicator is really the main way of interacting with the Indicators API, so to me it makes sense the keep the tests in the API spec, especially as the spec is pretty brief anyway.

akhenry commented:
@larkin Have made all requested changes [except one](https://github.com/nasa/openmct/pull/1837/files/a63403202334f947b6cf1c98998c48c20bc086db..35f57a706bc9f072a637dec09bfda5450cb46d39#r199288547) - happy to discuss this further. 

For convenience, [here are the changes made since the review](https://github.com/nasa/openmct/pull/1837/files/a63403202334f947b6cf1c98998c48c20bc086db..35f57a706bc9f072a637dec09bfda5450cb46d39).

## Author Checklist

* Changes address original issue? Y
* Unit tests included and/or updated with changes? Y
* Command line build passes? Y
* Changes have been smoke-tested? Y

akhenry commented:
> Can we move the "legacyIndicator" plugin to the "src/adapter" bundle? Everything else javascript-based looks good to me.
👍 Makes a lot more sense there, thanks.

charlesh88 commented:
@larkin As discussed:

1. Limits and warnings classes and approach needs a deeper discussion and more detailed documentation for users. I want to move that out of this Indicators PR and will file a new separate issue for us to track this need.
1. Will fix the missing s-status-on, my bad.
1. I'm going to revise the Indicators portion of the Style Guide to just focus on the usage of on/off/available/disabled states.
1. Agree on sync status, and think we should remove from the Style Guide for now.

charlesh88 commented:
@larkin @akhenry Changes pushed to `indicators-api` branch.

- I added more clarity and better examples in the Indicators section.
- `s-status-on` has been fixed.
- As mentioned above, limits and warnings approach needs a deeper discussion on the overall approach and more detailed documentation for users. As mentioned above, let's pick this up as we move deeper into the CSS and markup refactoring project in 2018. #1558 is a good starting point for this discussion.
- That said, I did do some refinements to content and examples in the Style Guide for both Status and Limits. If anything there doesn't seem right, do not hesitate to comment on that and I'll address.
- The "Sync" section has been removed from Guide.

larkin commented:
# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
